jormungandr/src/watch_client.rs

```diff
@@ -0,0 +1,359 @@
+        let (handle, future) = intercom::stream_reply(32);
+
+        self.request_tx
+            .clone()
+            .send(RequestMsg::SyncMultiverse { from, handle })
+            .await
+            .map_err(|e| chain_network::error::Error::new(Code::Unavailable, e))?;
+
+        let stream = future
+            .await
+            .map_err(|e| chain_network::error::Error::new(Code::Internal, e))?;
+
+        Ok(stream.map_err(|e| chain_network::error::Error::new(Code::Internal, e)))
+    }
+}
+
+async fn handle_sync_multiverse(
+    checkpoints: BlockIds,
+    blockchain: &Blockchain,
+    storage: &Storage,
+    sink: &mut intercom::ReplyStreamSink<Block>,
+) -> Result<(), intercom::Error> {
+    let block0 = blockchain.block0();
+
+    let (checkpoints, lsb_length, lsb_id) = {
+        let mut filtered = Vec::with_capacity(checkpoints.len());
+
+        let mut min_length = u32::MAX;
+        let mut min_index = None;
+
+        for id_raw in checkpoints.iter() {
+            let id = HeaderHash::deserialize(id_raw.as_bytes())
+                .map_err(intercom::Error::invalid_argument)?;
+
+            // the checkpoint could be unknown to the node because it was part of a branch that
+            // didn't survive the selection, in that case, we just ignore it and let the client
+            // realize that it can forget about it.
+            if let Some(chain_length) = storage.get_chain_length(id) {
+                filtered.push((chain_length, id));
+
+                // keep track of the min length in order to find the (expected) lsb
+                if chain_length < min_length {
+                    min_length = chain_length;
+
+                    // we called `push` up just before, so `len()` should be >= 1, and this won't
+                    // overflow
+                    min_index.replace(filtered.len() - 1);
+                }
+            }
+        }
+
+        // min_index will be None if there were no checkpoints, or if none of them were known to
+        // the node.
+        let (lsb_length, lsb_id) = min_index
+            .map(|idx| filtered.swap_remove(idx))
+            .unwrap_or((0, *block0));
+
+        (filtered, lsb_length, lsb_id)
+    };
+
+    // we are adding 1 to the lsb later, because if it is present then we don't need to send it.
+    // but if the checkpoints are empty then it means the last stable block is the block0, and we
+    // send it here.
+    if checkpoints.is_empty() {
+        let block = storage
+            .get(*block0)
+            .map_err(intercom::Error::failed)
+            .and_then(|maybe_block0| {
+                maybe_block0.ok_or_else(|| intercom::Error::failed("block0 not found in storage"))
+            })?;
+
+        sink.send(Ok(chain_network::data::Block::from_bytes(
+            block.serialize_as_vec().unwrap(),
+        )))
+        .await
+        .map_err(intercom::Error::failed)?;
+    }
```

Comment:
nit: you aren't actually accessing `lsb_id`/`lsb_length` here, so some of the logic could be simplified to just look at `Some`/`None`